docs/shared/concepts/customizing-inputs.md

```diff
@@ -0,0 +1,128 @@
+
+Knowing the syntax doesn't always explain how you would use the feature, so here are two scenarios explaining how to customize `inputs` to match your particular set up.
+
+## Scenario 1: React App
+
+Let's say we have a repo with two projects. There's a React app (`react-app`) that depends on a React UI library (`ui-library`).
+
+### Global Settings
+
+In the root `nx.json` file, we can set the following `namedInputs`:
+
+```jsonc
+{
+  "namedInputs": {
+    "sharedGlobals": ["{workspaceRoot}/babel.config.json"],
+    "default": ["{projectRoot}/**/*", "sharedGlobals"],
+    "production": [
+      "default",
+      "!{projectRoot}/**/*.spec.ts",
+      "!{projectRoot}/**/*.md",
+      "!{projectRoot}/tsconfig.spec.json",
+      "!{projectRoot}/jest.config.ts",
+      "!{projectRoot}/.eslintrc.json"
+    ]
+  }
+}
+```
+
+This configuration defines three `namedInputs`.
+
+- `sharedGlobals` contains root level files that should invalidate the cache for all projects. Note that `tsconfig.base.json` and the `package.json` lock files are handled behind the scenes by Nx so that when you modify the specific properties that affect your project, the cache is invalidated.
+- `default` contains every file in the project and everything defined in `sharedGlobals`
+- `production` starts with every file in `default`, but removes `.spec.ts` files, markdown files and some project-specific config files.
+
+These names are not hard-coded. You can add your own `namedInputs` or rename these as you see fit.
+
+With these `namedInputs` in place, we can set default `inputs` for project targets in `nx.json`:
+
+```jsonc
+{
+  "targetDefaults": {
+    "build": {
+      "inputs": ["production", "^production"],
+      "dependsOn": ["^build"]
+    },
+    "test": {
+      "inputs": ["default", "^production", "{workspaceRoot}/jest.preset.js"]
+    }
+  }
+}
+```
+
+In this configuration, `build` needs to be re-run if the project's `production` files are changed or the dependencies' `production` files are changed. The `"dependsOn": ["^build"]` setting ensures that the output files of the dependencies are correct, but it does not tell Nx to re-run this project's build target when those outputs change. That's why `"^production"` is needed in the `inputs` array.
+
+Note that we are using `^production` instead of `^default` because changing test files of a dependency should not invalidate the build cache for this project.
+
+The `test` target cache is invalidated if test or production code is changed in this project (`default`) or if production code is changed in a dependency `^production`. The tests also need to be run again if the root level `jest.preset.js` file is modified.
+
+### Project Settings
+
+In the `ui-library` library, we have some `*.spec.tsx` files that we don't want to count as production files. We can modify the project configuration to account for this:
+
+```jsonc
+{
```

Comment:
can we in some way add contextual information of what file this config is in? Like for all our code examples? Adding a top-level comment is one way, but I know that Docusaurus for instance allows you to specify a special property just after the syntax definition and that'd be included as a special header of the code-block rendering.

@bcabanes ?

Reply:
Like having the name of the file as a tab on top of the code?
Yeah, that's possible, I'll have to update the component to be able to do it.